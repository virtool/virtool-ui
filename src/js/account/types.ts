import { GroupMinimal, Permissions } from "@groups/types";
import { User } from "@users/types";

export enum QuickAnalyzeWorkflow {
    aodp = "aodp",
    nuvs = "nuvs",
    pathoscope_bowtie = "pathoscope_bowtie",
}

export type APIKeyMinimal = {
    created_at?: string;
    groups: Array<GroupMinimal>;
    id: string;
    key?: string;
    name: string;
    permissions: Permissions;
};

export type AccountSettings = {
    quick_analyze_workflow: QuickAnalyzeWorkflow;
    show_ids: boolean;
    show_versions: boolean;
    skip_quick_analyze_dialog: boolean;
};

export type Account = User & {
    settings: AccountSettings;
    email?: string;
};
