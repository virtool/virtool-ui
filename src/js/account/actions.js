/**
 * Redux action creators for use with the active user account.
 *
 * @module account/actions
 */
import { GET_ACCOUNT, LOGIN, RESET_PASSWORD } from "@app/actionTypes";
import { createAction } from "@reduxjs/toolkit";

/**
 * Returns action that can trigger an API call for getting the current account data.
 *
 * @func
 * @returns {object}
 */
export const getAccount = createAction(GET_ACCOUNT.REQUESTED);

/**
 * Returns action that can trigger an API call that will login into a new session.
 *
 * @func
 * @returns {object}
 */
export const login = createAction(LOGIN.REQUESTED, (username, password, remember) => ({
    payload: { username, password, remember },
}));

/**
 * Returns action that updates the redux store to assume the user is logged in
 *
 * @func
 * @returns {object}
 */
export const loginSucceeded = createAction(LOGIN.SUCCEEDED, () => ({
    payload: {
        reset: false,
    },
}));

/**
 * Returns action that can trigger an API call that will reset the password for the user associated with the  current
 * session.
 *
 * @func
 * @returns {object}
 */
export const resetPassword = createAction(RESET_PASSWORD.REQUESTED, (password, resetCode) => ({
    payload: { password, resetCode },
}));
