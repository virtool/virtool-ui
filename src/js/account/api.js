/**
 * Defines methods used for calling account API endpoints.
 *
 * @module account/api
 */

import { Request } from "../app/request";

/**
 * Gets the complete data for the current account.
 *
 * @func
 * @returns {promise}
 */
export const get = () => Request.get("/account");

/**
 * Updates the complete data for the current account.
 *
 * @func
 * @param update {object} the update to apply to current account.
 * @returns {promise}
 */
export const update = ({ update }) => Request.patch("/account").send(update);

/**
 * Gets the settings object for the current account.
 *
 * @func
 * @returns {promise}
 */
export const getSettings = () => Request.get("/account/settings");

/**
 * Updates the settings for the current account.
 *
 * @func
 * @param update {object} the update to apply to account settings
 * @returns {promise}
 */
export const updateSettings = ({ update }) => Request.patch("/account/settings").send(update);

/**
 * Changes the password for the current account.
 *
 * @func
 * @param oldPassword {string} the old password (for verification)
 * @param newPassword {string} the new password
 * @returns {promise}
 */
export const changePassword = ({ old_password, password }) =>
    Request.patch("/account").send({
        old_password,
        password,
    });

/**
 * Gets all API keys owned by the current account.
 *
 * @returns {promise}
 */
export const getAPIKeys = () => Request.get("/account/keys");

/**
 * Create a new API key for the current account.
 *
 * @func
 * @param name {string} a name for the API key
 * @param permissions {object} the permissions to assign to the API key
 * @returns {promise}
 */
export const createAPIKey = ({ name, permissions }) =>
    Request.post("/account/keys").send({
        name,
        permissions,
    });

/**
 * Update the permissions for an existing API key owned by the current account.
 *
 * @func
 * @param keyId {string} the unique id for the API key to update
 * @param permissions {object} the new permissions for the API key
 * @returns {promise}
 */
export const updateAPIKey = ({ keyId, permissions }) =>
    Request.patch(`/account/keys/${keyId}`).send({
        permissions,
    });

/**
 * Remove an existing API key owned by current account.
 *
 * @param keyId
 * @returns {promise}
 */
export const removeAPIKey = ({ keyId }) => Request.delete(`/account/keys/${keyId}`);

export const login = ({ username, password, remember }) =>
    Request.post("/account/login").send({ username, password, remember });

/**
 * Logs out the current session.
 *
 * @func
 * @returns {promise}
 */
export const logout = () => Request.get("/account/logout");

export const resetPassword = ({ password, resetCode }) =>
    Request.post("/account/reset").send({ password, reset_code: resetCode });
