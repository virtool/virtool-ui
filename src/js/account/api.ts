/**
 * Functions for requesting account data for the current user from backend
 *
 * @module account/api
 */

import { Request } from "@app/request";
import { Permissions } from "@groups/types";
import { User } from "@users/types";
import { Response } from "superagent";
import { Account, APIKeyMinimal } from "./types";

/**
 * Gets complete account data for the current user.
 *
 * @returns A promise resolving to a response containing the
 * current user's account data.
 */
export function get(): Promise<Response> {
    return Request.get("/account");
}

export type AccountUpdate = {
    email?: string;
};

/**
 * Updates the complete data for the current account.
 *
 * @param update - The update to apply to current account.
 * @returns A promise resolving to a response containing the updated
 * user's account data
 */
export function updateAccount(update: AccountUpdate): Promise<User> {
    return Request.patch("/account")
        .send({ update })
        .then(res => res.body);
}

/**
 * Gets the settings object for the current account.
 *
 * @returns A promise resolving to a response containing the
 * current user's personal settings.
 */
export function getSettings(): Promise<Response> {
    return Request.get("/account/settings");
}

/**
 * Updates the settings for the current account.
 *
 * @param update - The update to apply to account settings
 * @returns A promise resolving to a response containing the
 * user's updated personal settings.
 */
export function updateSettings({ update }): Promise<Response> {
    return Request.patch("/account/settings").send(update);
}

/**
 * Changes the password for the current account.
 *
 * @param old_password - The old password (for verification)
 * @param password - The new password
 * @returns A promise resolving to a response indicating if the
 * password was successfully changed.
 */
export function changePassword(old_password: string, password: string): Promise<User> {
    return Request.patch("/account")
        .send({
            old_password,
            password,
        })
        .then(res => res.body);
}

/**
 * Gets all API keys owned by the current account.
 *
 * @returns A promise resolving to a response containing the
 * current user's API keys.
 */
export function getAPIKeys(): Promise<APIKeyMinimal[]> {
    return Request.get("/account/keys").then(res => res.body);
}

/**
 * Create a new API key for the current account.
 *
 * @param name - a name for the API key
 * @param permissions - Complete list of permissions for the API key
 * @returns A promise resolving to a response containing the newly created API key
 */
export function createAPIKey(name: string, permissions: Permissions): Promise<APIKeyMinimal> {
    return Request.post("/account/keys")
        .send({
            name,
            permissions,
        })
        .then(res => res.body);
}

/**
 * Update the permissions for an existing API key owned by the current account.
 *
 * @param keyId - The unique id for the API key to update
 * @param permissions - The new permissions for the API key
 * @returns A promise resolving to a response containing the updated API key
 */
export function updateAPIKey(keyId: string, permissions: Permissions): Promise<APIKeyMinimal> {
    return Request.patch(`/account/keys/${keyId}`)
        .send({
            permissions,
        })
        .then(res => res.body);
}

/**
 * Remove an existing API key owned by current account.
 *
 * @param keyId - The unique id of the API key to remove
 * @returns A promise which resolves to a response indicating if the API key was successfully removed
 */
export function removeAPIKey(keyId: string): Promise<null> {
    return Request.delete(`/account/keys/${keyId}`).then(res => res.body);
}

/**
 * Log in using the provided credentials.
 *
 * @param username - The username to log in with
 * @param password - The password to log in with
 * @param remember - Whether the sessions should be remembered for a
 * longer period of time
 * @returns A promise which resolves to a response indicating if the users
 * password must be reset and required information if it needs to be.
 */
export function login({
    username,
    password,
    remember,
}: {
    username: string;
    password: string;
    remember: boolean;
}): Promise<Response> {
    return Request.post("/account/login").send({ username, password, remember });
}

/**
 * Logs out the current session.
 *
 * @returns A promise which resolves to a response indicating if the
 * logout was successful.
 */
export function logout(): Promise<null> {
    return Request.get("/account/logout").then(res => res.body);
}

/**
 * Reset the current users password
 *
 * @param password - The new password
 * @param resetCode - The reset code sent to the user
 * @returns A promise which resolves to a response indicating if the
 * password was successfully reset.
 */
export function resetPassword({ password, resetCode }: { password: string; resetCode: string }): Promise<Response> {
    return Request.post("/account/reset").send({ password, reset_code: resetCode });
}

/**
 * Fetches account data for the logged-in user
 *
 * @returns A Promise resolving to the current user's account data
 */
export function fetchAccount(): Promise<Account> {
    return Request.get("/account")
        .query()
        .then(response => response.body);
}
