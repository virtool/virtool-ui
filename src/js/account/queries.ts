import { ErrorResponse } from "@/types/types";
import { Permissions } from "@groups/types";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { User } from "@users/types";
import { resetClient } from "@utils/utils";
import {
    AccountUpdate,
    changePassword,
    createAPIKey,
    fetchAccount,
    getAPIKeys,
    logout,
    removeAPIKey,
    updateAccount,
    updateAPIKey,
} from "./api";
import { Account, APIKeyMinimal } from "./types";

/**
 * Factory object for generating account query keys
 */
export const accountKeys = {
    all: () => ["account"],
    details: () => ["account", "details"] as const,
};

/**
 * Fetches account data for the logged-in user
 *
 * @returns UseQueryResult object containing the account data
 */
export function useFetchAccount() {
    return useQuery<Account>(accountKeys.all(), () => fetchAccount());
}

/**
 * Initializes a mutator for updating a user
 *
 * @returns A mutator for updating a user
 */
export function useUpdateAccount() {
    const queryClient = useQueryClient();

    return useMutation<User, ErrorResponse, { update: AccountUpdate }>(({ update }) => updateAccount(update), {
        onSuccess: () => {
            queryClient.invalidateQueries(accountKeys.all());
        },
    });
}

/**
 * Initializes a mutator for updating a user
 *
 * @returns A mutator for updating a user
 */
export function useChangePassword() {
    const queryClient = useQueryClient();

    return useMutation<User, ErrorResponse, { old_password: string; password: string }>(
        ({ old_password, password }) => changePassword(old_password, password),
        {
            onSuccess: () => {
                queryClient.invalidateQueries(accountKeys.all());
            },
        },
    );
}

/**
 * Fetches the API keys for the current user
 *
 * @returns A list of API keys for the current user
 */
export function useFetchAPIKeys() {
    return useQuery<APIKeyMinimal[]>(accountKeys.details(), () => getAPIKeys());
}

/**
 * Initializes a mutator for creating a new API key
 *
 * @returns A mutator for creating a new API key
 */
export function useCreateAPIKey() {
    return useMutation<APIKeyMinimal, ErrorResponse, { name: string; permissions: Permissions }>(
        ({ name, permissions }) => createAPIKey(name, permissions),
    );
}

/**
 * Initializes a mutator for updating an API key
 *
 * @returns A mutator for updating an API key
 */
export function useUpdateAPIKey() {
    const queryClient = useQueryClient();

    return useMutation<APIKeyMinimal, ErrorResponse, { keyId: string; permissions: Permissions }>(
        ({ keyId, permissions }) => updateAPIKey(keyId, permissions),
        {
            onSuccess: () => {
                queryClient.invalidateQueries(accountKeys.all());
            },
        },
    );
}

/**
 * Initializes a mutator for removing an API key
 *
 * @returns A mutator for removing an API key
 */
export function useRemoveAPIKey() {
    const queryClient = useQueryClient();

    return useMutation<null, ErrorResponse, { keyId: string }>(({ keyId }) => removeAPIKey(keyId), {
        onSuccess: () => {
            queryClient.invalidateQueries(accountKeys.all());
        },
    });
}

/**
 * Initializes a mutator for logging out a user
 *
 * @returns A mutator for logging out a user
 */
export function useLogout() {
    return useMutation<null, ErrorResponse>(logout, {
        onSuccess: () => {
            resetClient();
        },
    });
}
