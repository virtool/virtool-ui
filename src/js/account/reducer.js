/**
 * Exports a reducer for dealing with the account of the session user.
 *
 * @module account/reducer
 */
import { createReducer } from "@reduxjs/toolkit";
import {
    CLEAR_API_KEY,
    CREATE_API_KEY,
    GET_ACCOUNT,
    GET_API_KEYS,
    LOGOUT,
    UPDATE_ACCOUNT,
    UPDATE_ACCOUNT_SETTINGS
} from "../app/actionTypes";

/**
 * The state that should initially be stored.
 *
 * @const
 * @type {object}
 */
export const initialState = {
    apiKeys: null,
    newKey: null,
    ready: false
};

/**
 * A reducer for dealing with the account of the session user.
 *
 * @param state {object}
 * @param action {object}
 * @returns {object}
 */
export const accountReducer = createReducer(initialState, builder => {
    builder
        .addCase(GET_ACCOUNT.SUCCEEDED, (state, action) => {
            return { ...state, ...action.payload, ready: true };
        })
        .addCase(UPDATE_ACCOUNT.SUCCEEDED, (state, action) => {
            return { ...state, ...action.payload };
        })
        .addCase(GET_API_KEYS.SUCCEEDED, (state, action) => {
            state.apiKeys = action.payload;
        })
        .addCase(CREATE_API_KEY.REQUESTED, state => {
            state.key = null;
        })
        .addCase(CREATE_API_KEY.SUCCEEDED, (state, action) => {
            state.newKey = action.payload.key;
        })
        .addCase(CLEAR_API_KEY, state => {
            state.newKey = null;
        })
        .addCase(UPDATE_ACCOUNT_SETTINGS.SUCCEEDED, (state, action) => {
            state.settings = action.payload;
        })
        .addCase(LOGOUT.SUCCEEDED, () => {
            return { ...initialState };
        });
});

export default accountReducer;
