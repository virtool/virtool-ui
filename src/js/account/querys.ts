import { useQuery } from "@tanstack/react-query";
import { fetchAccount } from "./api";
import { Account } from "./types";

/**
 * Factory object for generating account query keys
 */
export const accountKeys = {
    all: () => ["account"],
};

/**
 * Fetches account data for the logged-in user
 *
 * @returns UseQueryResult object containing the account data
 */
export const useFetchAccount = () => {
    return useQuery<Account>(accountKeys.all(), () => fetchAccount());
};
