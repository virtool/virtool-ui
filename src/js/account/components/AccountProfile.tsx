import AccountGroups from "@account/components/AccountGroups";
import { useFetchAccount } from "@account/queries";
import { getFontSize, getFontWeight } from "@app/theme";
import { Icon, InitialIcon, Label, LoadingPlaceholder } from "@base";
import React from "react";
import styled from "styled-components";
import ChangePassword from "./ChangePassword";
import Email from "./Email";

const AccountProfileHeader = styled.div`
    align-items: center;
    display: flex;
    margin-bottom: 15px;
    > div {
        flex: 2 0 auto;
        margin-left: 15px;

        h3 {
            align-items: center;
            display: flex;
            justify-content: space-between;
            flex: 2 0 auto;
            font-size: ${getFontSize("xl")};
            font-weight: ${getFontWeight("thick")};
            line-height: 1.2;
            margin: 0;
        }
    }
`;

/**
 * Displays information related to the users account with options to reset password and email
 */
export default function AccountProfile() {
    const { data, isLoading } = useFetchAccount();

    if (isLoading) {
        return <LoadingPlaceholder />;
    }

    const { administrator_role, email, groups, handle, last_password_change } = data;

    return (
        <>
            <AccountProfileHeader>
                <InitialIcon handle={handle} size="xxl" />
                <div>
                    <h3>
                        {handle}
                        {administrator_role && (
                            <Label key="administrator" className="capitalize text-base ml-auto" color="purple">
                                <Icon name="user-shield" /> {administrator_role} Administrator
                            </Label>
                        )}
                    </h3>
                </div>
            </AccountProfileHeader>

            <ChangePassword lastPasswordChange={last_password_change} />
            <Email email={email} />
            <AccountGroups groups={groups} />
        </>
    );
}
