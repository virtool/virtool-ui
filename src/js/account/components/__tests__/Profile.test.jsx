import { AdministratorRoles } from "@administration/types";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { connectRouter } from "connected-react-router";
import { createBrowserHistory } from "history";
import React from "react";
import { combineReducers } from "redux";
import { beforeEach, describe, expect, it } from "vitest";
import { createFakeAccount } from "../../../../tests/fake/account";
import { createGenericReducer, renderWithRouter } from "../../../../tests/setupTests";
import AccountProfile from "../Profile";

function createReducer(state, history) {
    return combineReducers({
        account: createGenericReducer(state.account),
        settings: createGenericReducer(state.settings),
        router: connectRouter(history),
    });
}

describe("<AccountProfile />", () => {
    let history;
    let state;

    beforeEach(() => {
        history = createBrowserHistory();

        state = {
            account: { ...createFakeAccount(), administrator_role: AdministratorRoles.FULL, handle: "amanda36" },
            settings: { data: { minimum_password_length: 8 } },
        };
    });

    it("should render when administrator", () => {
        renderWithRouter(<AccountProfile />, state, history, createReducer);

        expect(screen.getByText("amanda36")).toBeInTheDocument();
        expect(screen.getByText("full Administrator")).toBeInTheDocument();
        expect(screen.queryAllByText("AM")).toHaveLength(2);
    });

    it("should render when not administrator", () => {
        state.account.administrator_role = null;
        renderWithRouter(<AccountProfile />, state, history, createReducer);

        expect(screen.getByText("amanda36")).toBeInTheDocument();
    });

    it("should render with initial email", () => {
        state.account.email = "virtool.devs@gmail.com";
        renderWithRouter(<AccountProfile />, state, history, createReducer);
        const emailInput = screen.getByLabelText("Email Address");
        expect(emailInput.value).toBe("virtool.devs@gmail.com");
    });

    it("should handle email changes", async () => {
        state.account.email = "";
        renderWithRouter(<AccountProfile />, state, history, createReducer);

        const emailInput = screen.getByLabelText("Email Address");
        expect(emailInput.value).toBe("");

        await userEvent.type(emailInput, "invalid");
        await userEvent.click(screen.getByText("Save"), { role: "button" });

        expect(emailInput).toHaveValue("invalid");
        expect(screen.getByText("Please provide a valid email address")).toBeInTheDocument();

        await userEvent.clear(emailInput);
        await userEvent.type(emailInput, "virtool.devs@gmail.com");
        await userEvent.click(screen.getByText("Save"), { role: "button" });
    });

    it("should handle password changes", async () => {
        renderWithRouter(<AccountProfile />, state, history, createReducer);

        const oldPasswordInput = screen.getByLabelText("Old Password");
        const newPasswordInput = screen.getByLabelText("New Password");

        // Try without providing old password.
        await userEvent.type(newPasswordInput, "long_enough_password");
        await userEvent.click(screen.getByText("Change"), { role: "button" });

        expect(screen.getByText("Please provide your old password")).toBeInTheDocument();

        await userEvent.clear(newPasswordInput);
        await userEvent.type(oldPasswordInput, "expected_password");
        await userEvent.type(newPasswordInput, "short");

        expect(screen.getByLabelText("New Password")).toHaveValue("short");

        await userEvent.click(screen.getByText("Change"), { role: "button" });

        expect(screen.getByText("Passwords must contain at least 8 characters")).toBeInTheDocument();
    });
});
