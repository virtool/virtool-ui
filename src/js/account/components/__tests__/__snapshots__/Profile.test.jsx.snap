// Vitest Snapshot v1

exports[`<AccountProfile /> > should render when administrator 1`] = `
<div>
  <styled.div>
    <InitialIcon
      handle="foo"
      size="xxl"
    />
    <div>
      <h3>
        foo
        <styled.span
          color="purple"
          key="administrator"
        >
          <Icon
            faStyle="fas"
            fixedWidth={false}
            name="user-shield"
          />
           Administrator
        </styled.span>
      </h3>
      <styled.div>
        <styled.span
          key="ac091sa1"
        >
          <Icon
            faStyle="fas"
            fixedWidth={false}
            name="users"
          />
           
          Technicians
        </styled.span>
        <styled.span
          key="980s921a"
        >
          <Icon
            faStyle="fas"
            fixedWidth={false}
            name="users"
          />
           
          Managers
        </styled.span>
      </styled.div>
    </div>
  </styled.div>
  <Connect(Email) />
  <Connect(ChangePassword) />
</div>
`;

exports[`<AccountProfile /> > should render when not administrator 1`] = `
<div>
  <styled.div>
    <InitialIcon
      handle="foo"
      size="xxl"
    />
    <div>
      <h3>
        foo
      </h3>
      <styled.div>
        <styled.span
          key="ac091sa1"
        >
          <Icon
            faStyle="fas"
            fixedWidth={false}
            name="users"
          />
           
          Technicians
        </styled.span>
        <styled.span
          key="980s921a"
        >
          <Icon
            faStyle="fas"
            fixedWidth={false}
            name="users"
          />
           
          Managers
        </styled.span>
      </styled.div>
    </div>
  </styled.div>
  <Connect(Email) />
  <Connect(ChangePassword) />
</div>
`;
